src/Tizen.MachineLearning.Inference/Tizen.MachineLearning.Inference/Pipeline.cs

```diff
@@ -0,0 +1,539 @@
+        public void SubscribeDataEvent(string name)
+        {
+            NNStreamer.CheckNNStreamerSupport();
+
+            /* Check the argument */
+            if (string.IsNullOrEmpty(name))
+                throw NNStreamerExceptionFactory.CreateException(NNStreamerError.InvalidParameter, "Node name is invalid");
+
+            if (!_sinkNodeList.ContainsKey(name))
+            {
+                SinkNode sink = new SinkNode(name, this);
+
+                /* Consider multiple sink nodes in pipeline */
+                _sinkNodeList.Add(name, sink);
+            }
+        }
+
+        /// <summary>
+        /// Unsubscribes data event from sink node.
+        /// </summary>
+        /// <param name="name">The name of sink node</param>
+        /// <feature>http://tizen.org/feature/machine_learning.inference</feature>
+        /// <exception cref="NotSupportedException">Thrown when the feature is not supported.</exception>
+        /// <exception cref="ArgumentException">Thrown when the method failed due to an invalid parameter.</exception>
+        /// <since_tizen> 8 </since_tizen>
+        public void UnsubscribeDataEvent(string name)
+        {
+            NNStreamer.CheckNNStreamerSupport();
+
+            /* Check the argument */
+            if (string.IsNullOrEmpty(name) || !_sinkNodeList.ContainsKey(name))
+                throw NNStreamerExceptionFactory.CreateException(NNStreamerError.InvalidParameter, "Node name is invalid");
+
+            SinkNode sink = _sinkNodeList[name];
+
+            sink.Unregister();
+            _sinkNodeList.Remove(name);
+        }
+
+        /// <summary>
+        /// Internal method to send data event from sink node.
+        /// </summary>
+        /// <param name="name">The name of sink node</param>
+        /// <param name="data">Newly received data</param>
+        /// <since_tizen> 8 </since_tizen>
+        internal void SendDataEvent(string name, TensorsData data)
+        {
+            _dataReceived?.Invoke(this, new DataReceivedEventArgs(name, data));
+        }
+
+        /// <summary>
+        /// Gets the source node instance with given node name.
+        /// </summary>
+        /// <param name="name">The name of source node</param>
+        /// <feature>http://tizen.org/feature/machine_learning.inference</feature>
+        /// <exception cref="NotSupportedException">Thrown when the feature is not supported.</exception>
+        /// <exception cref="ArgumentException">Thrown when the method failed due to an invalid parameter.</exception>
+        /// <since_tizen> 8 </since_tizen>
+        public SourceNode GetSource(string name)
+        {
+            NNStreamer.CheckNNStreamerSupport();
+
+            /* Check the parameter */
+            if (string.IsNullOrEmpty(name))
+                throw NNStreamerExceptionFactory.CreateException(NNStreamerError.InvalidParameter, "Node name is invalid");
+
+            SourceNode node = new SourceNode(name, this);
```

Comment:
How about apply cache? because usually app developer will directly use a return value like below
``` c#
 pipeline.GetSource("src1").Input(data);
```
in this case, will be created many internal handles on native.

Reply:
I just updated this PR. (add NodeInfo class for common node information)
Please review the changes.